server/supernova/audio_backend/portaudio_backend.hpp

```diff
@@ -121,15 +101,43 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
         }
     }
 
-    bool open_stream(std::string const& input_device, unsigned int inchans, std::string const& output_device,
-                     unsigned int outchans, unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
+    bool open_stream(std::string& input_device, unsigned int inchans, std::string& output_device, unsigned int outchans,
```

Comment:
just so i'm clear on semantics -- if input_device and output_device are empty and inchans and outchans are nonzero, input_device and output_device will be overwritten? or are there other cases where these values will be changed? this is not clear to me and the documentation for TryGetDefaultPaDevices doesn't help; perhaps you could update the documentation for this function or that one to make this easier to understand

Reply:
You are correct, if `input_device` and `output_device` are empty (and `inchans`/`outchans` are nonzero), they will be overwritten with the name of the default device that `TryGetDefaultPaDevices` is setting. AFAIU setting these strings here is only used for posting purposes in `main.cpp` after opening the stream, and only when using default device selection (otherwise these strings are set from argv and not altered insude `open_stream`).
Previously, default device selection was happening in `main.cpp` before `open_stream` and there was no need to set these strings inside `open_stream`. If this is weird now, I could move status posting inside `open_stream` and remove setting these strings / using them after `open_stream` in `main.cpp`. Do you think that would be clearer?

Reply:
as a side note, it is unclear to me why we have `ifdef`s in `main.cpp` for different backends, when it seems like `start_audio_backend` could be moved to individual backends` header files? Though I have not investigated this logic fully so maybe there's a legitimate reason for it.